File: src/fsharp/examples/cublas/Gemm.fs
(*** hide ***)
module Tutorial.Fs.examples.cublas.Gemm

open System
open Microsoft.FSharp.Quotations
open Alea.CUDA
open Alea.CUDA.Utilities
open Alea.CUDA.CULib
open NUnit.Framework
open NUnit.Framework.Constraints
open FsUnit

(**
Matrix multiplication with `<t>gemm`. 
*)

let cublas = CUBLAS.Default

(*** define:gemmCpu ***)
module cpu =     
    let dgemm transa transb m n k alpha (A:float[]) lda (B:float[]) ldb beta (C:float[]) ldc =
        let A = Array2D.ofArrayColumnMajor lda k A
        let B = Array2D.ofArrayColumnMajor ldb n B
        let C = Array2D.ofArrayColumnMajor ldc n C
        for j in [1..n] do
            for i in [1..m] do
                C.[i - 1,j - 1] <- beta * C.[i - 1,j - 1]
            for l in [1..k] do
                let temp = alpha * B.[l - 1,j - 1]
                for i in [1..m] do
                    C.[i - 1,j - 1] <- C.[i - 1,j - 1] + temp * A.[i - 1,l - 1]
        C |> Array2D.toArrayColumnMajor

    let zgemm transa transb m n k (alpha:double2) (A:double2[]) lda (B:double2[]) ldb (beta:double2) (C:double2[]) ldc =
        let A = double2.ToComplex(A) |> Array2D.ofArrayColumnMajor lda k 
        let B = double2.ToComplex(B) |> Array2D.ofArrayColumnMajor ldb n 
        let C = double2.ToComplex(C) |> Array2D.ofArrayColumnMajor ldc n
        for j in [1..n] do
            for i in [1..m] do
                C.[i - 1,j - 1] <- beta.ToComplex() * C.[i - 1,j - 1]
            for l in [1..k] do
                let temp = alpha.ToComplex() * B.[l - 1,j - 1]
                for i in [1..m] do
                    C.[i - 1,j - 1] <- C.[i - 1,j - 1] + temp * A.[i - 1,l - 1]
        C |> Array2D.toArrayColumnMajor |> double2.OfComplex
        
(*** define:gemmGpu ***)
module gpu =    
    let dgemm transa transb m n k alpha (A:float[]) lda (B:float[]) ldb beta (C:float[]) ldc =
        let worker = Worker.Default
        use dalpha = worker.Malloc([|alpha|])
        use dA = worker.Malloc(A)
        use dB = worker.Malloc(B)
        use dbeta = worker.Malloc([|beta|])
        use dC = worker.Malloc(C)
        cublas.Dgemm(transa, transb, m, n, k, dalpha.Ptr, dA.Ptr, lda, dB.Ptr, ldb, dbeta.Ptr, dC.Ptr, ldc)
        dC.Gather()

    let zgemm transa transb m n k alpha (A:double2[]) lda (B:double2[]) ldb beta (C:double2[]) ldc =
        let worker = Worker.Default
        use dalpha = worker.Malloc([|alpha|])
        use dA = worker.Malloc(A)
        use dB = worker.Malloc(B)
        use dbeta = worker.Malloc([|beta|])
        use dC = worker.Malloc(C)
        cublas.Zgemm(transa, transb, m, n, k, dalpha.Ptr, dA.Ptr, lda, dB.Ptr, ldb, dbeta.Ptr, dC.Ptr, ldc)
        dC.Gather()

(*** define:gemmTest ***)
[<Test>]
let dgemmTest() =
    let gen rows cols = 
        Array.init (rows*cols) (TestUtil.genRandomDouble -5.0 5.0)
        
    let transa = cublasOperation_t.CUBLAS_OP_N
    let transb = cublasOperation_t.CUBLAS_OP_N
    let m,n,k = 5,5,5
    let lda = m // lda >= max(1,m)
    let ldb = n // ldb >= max(1,k)
    let ldc = k // ldc >= max(1,m)
    
    let alpha, beta = 1.5, 0.5
    
    let A = gen lda k
    let B = gen ldb n
    let C = gen ldc n

    let outputs = gpu.dgemm transa transb m n k alpha A lda B ldb beta C ldc
    let expected = cpu.dgemm transa transb m n k alpha A lda B ldb beta C ldc

    printfn "cpu result: %A" expected
    printfn "gpu result: %A" outputs

    outputs |> should (equalWithin 1e-12) expected 

[<Test>]
let zgemmTest() =
    let gen rows cols = 
        Array.init (rows*cols) (TestUtil.genRandomDouble2 -5.0 5.0)
        
    let transa = cublasOperation_t.CUBLAS_OP_N
    let transb = cublasOperation_t.CUBLAS_OP_N
    let m,n,k = 5,5,5
    let lda = m // lda >= max(1,m)
    let ldb = n // ldb >= max(1,k)
    let ldc = k // ldc >= max(1,m)
    
    let alpha = double2(2.0, 0.5)
    let beta = double2(1.5, 0.5)
    
    let A = gen lda k
    let B = gen ldb n
    let C = gen ldc n

    let outputs = gpu.zgemm transa transb m n k alpha A lda B ldb beta C ldc
    let expected = cpu.zgemm transa transb m n k alpha A lda B ldb beta C ldc

    (outputs, expected)
    ||> Array.iter2 (fun o e -> 
        o.x |> should (equalWithin 1e-9) e.x
        o.y |> should (equalWithin 1e-9) e.y)

    printfn "cpu result: %A" expected
    printfn "gpu result: %A" outputs

File: src/csharp/examples/cublas/Gemm.cs
using System;
using System.Numerics;
using Alea.CUDA;
using Alea.CUDA.CULib;
using Alea.CUDA.Utilities;
using NUnit.Framework;
using cublasOperation_t = Alea.CUDA.CULib.CUBLASInterop.cublasOperation_t;

namespace Tutorial.Cs.examples.cublas
{
    class Gemm
    {
        private static readonly CUBLAS Cublas = CUBLAS.Default;
        private static readonly Worker Worker = Cublas.Worker;

        private const cublasOperation_t Transa = cublasOperation_t.CUBLAS_OP_N;
        private const cublasOperation_t Transb = cublasOperation_t.CUBLAS_OP_N;
        
        //[gemmCpu]
        public static class cpu
        {
            public static double[] Dgemm(int m, int n, int k, double alpha, double[] _A, int lda, double[] _B, int ldb, double beta, double[] _C, int ldc)
            {
                var A = Array2D.ofArrayColumnMajor(lda, k, _A);
                var B = Array2D.ofArrayColumnMajor(ldb, n, _B);
                var C = Array2D.ofArrayColumnMajor(ldc, n, _C);

                for (var j = 1; j <= n; j++)
                {
                    for (var i = 1; i <= m; i++)
                        C[i - 1, j - 1] = beta * C[i - 1, j - 1];

                    for (var l = 1; l <= k; l++)
                    {
                        var temp = alpha * B[l - 1, j - 1];
                        for (var i = 1; i <= m; i++)
                            C[i - 1, j - 1] += temp * A[i - 1, l - 1];
                    }
                }
                return Array2D.toArrayColumnMajor(C);
            }
            
            public static double2[] Zgemm(int m, int n, int k, double2 alpha, double2[] _A, int lda, double2[] _B, int ldb, double2 beta, double2[] _C, int ldc)
            {
                var A = Array2D.ofArrayColumnMajor(lda, k, double2.ToComplex(_A));
                var B = Array2D.ofArrayColumnMajor(ldb, n, double2.ToComplex(_B));
                var C = Array2D.ofArrayColumnMajor(ldc, n, double2.ToComplex(_C));
                var R = Array2D.ofArrayColumnMajor(ldc, n, double2.ToComplex(_C));
                for (var j = 1; j <= n; ++j)
                {
                    for (var i = 1; i <= m; ++i)
                        R[i - 1, j - 1] = beta.ToComplex() * C[i - 1, j - 1];
                    for (var l = 1; l <= k; ++l)
                    {
                        var temp = alpha.ToComplex() * B[l - 1, j - 1];
                        for (var i = 1; i <= m; ++i)
                            R[i - 1, j - 1] += temp * A[i - 1, j - 1];
                    }

                }
                return double2.OfComplex(Array2D.toArrayColumnMajor(R));
            }
        }
        //[/gemmCpu]

        //[gemmGpu]
        public static class gpu
        {
            public static double[] Dgemm(int m, int n, int k, double alpha, double[] A, int lda, double[] B, int ldb, double beta, double[] C, int ldc)
            {
                using (var dA = Worker.Malloc(A))
                using (var dB = Worker.Malloc(B))
                using (var dC = Worker.Malloc(C))
                using (var dAlpha = Worker.Malloc(new[] { alpha }))
                using (var dBeta = Worker.Malloc(new[] { beta }))
                {
                    Cublas.Dgemm(Transa, Transb, m, n, k, dAlpha.Ptr, dA.Ptr, lda, dB.Ptr, ldb, dBeta.Ptr, dC.Ptr, ldc);
                    return dC.Gather();
                }
            }

            public static double2[] Zgemm(int m, int n, int k, double2 alpha, double2[] A, int lda, double2[] B, int ldb, double2 beta, double2[] C, int ldc)
            {
                using (var dA = Worker.Malloc(A))
                using (var dB = Worker.Malloc(B))
                using (var dC = Worker.Malloc(C))
                using (var dAlpha = Worker.Malloc(new[] { alpha }))
                using (var dBeta = Worker.Malloc(new[] { beta }))
                {
                    Cublas.Zgemm(Transa, Transb, m, n, k, dAlpha.Ptr, dA.Ptr, lda, dB.Ptr, ldb, dBeta.Ptr, dC.Ptr, ldc);
                    return dC.Gather();
                }
            }
        }
        //[/gemmGpu]
    }
    
    //[gemmTest]
    public partial class Test
    {
        private const int M = 5;
        private const int N = 5;
        private const int K = 5;

        private const int Lda = M; // lda >= max(1,m)
        private const int Ldb = N; // ldb >= max(1,k)
        private const int Ldc = K; // ldc >= max(1,m)

        [Test]
        public static void DgemmTest()
        {
            var rng = new Random();

            const double alpha = 2.0;
            const double beta = 1.0;

            var A = new double[Lda*K];
            var B = new double[Ldb*N];
            var C = new double[Ldc*N];

            for (var i = 0; i < Lda*K; ++i)
                A[i] = rng.NextDouble();

            for (var i = 0; i < Lda*N; ++i)
            {
                B[i] = rng.NextDouble();
                C[i] = rng.NextDouble();
            }
            
            var outputs = Gemm.gpu.Dgemm(M, N, K, alpha, A, Lda, B, Ldb, beta, C, Ldc);
            var expected = Gemm.cpu.Dgemm(M, N, K, alpha, A, Lda, B, Ldb, beta, C, Ldc);

            for (var i = 0; i < outputs.Length; ++i)
                Assert.AreEqual(outputs[i], expected[i], 1e-12);
        }

        [Test]
        public static void ZgemmTest()
        {
            var alpha = new double2(2.0, 0.5);
            var beta = new double2(1.0, 0.5);

            var A = new double2[Lda*K];
            var B = new double2[Ldb*N];
            var C = new double2[Ldc*N];

            var outputs = Gemm.gpu.Zgemm(M, N, K, alpha, A, Lda, B, Ldb, beta, C, Ldc);
            var expected = Gemm.cpu.Zgemm(M, N, K, alpha, A, Lda, B, Ldb, beta, C, Ldc);

            for (var i = 0; i < outputs.Length; ++i)
            {
                Assert.AreEqual(outputs[i].x, expected[i].x, 1e-12);
                Assert.AreEqual(outputs[i].y, expected[i].y, 1e-12);
            }
        }
    }
    //[/gemmTest]
}
